### PR TITLE
Reduce trivial warning messages from `tests/sampler_tests`

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

A part of #3815. 

## Description of the changes

- Resolve `step` related warning message such as `suggest_int(low=1, high=10, step=2`)
- Ignore the expected `UserWarning`
- Resolve numpy warning message via TPE (see the following my comment too)

Note that there are still several warning messages from TPE sampler and BoTorchSampler, I don't remove such messages because I suppose such changes make this PR more complicated.

### CODE REVIEW

File: optuna/samplers/_tpe/sampler.py [hunk 1]
```diff
@@ -714,7 +714,7 @@ def _split_observation_pairs(
             (infeasible_idx,) = (violation_1d > 0).nonzero()
             assert len(feasible_idx) >= n_below
             feasible_below, feasible_above = _split_observation_pairs(
-                list(np.array(loss_vals)[feasible_idx]), n_below, None
+                [loss_vals[i] for i in feasible_idx], n_below, None
```
Comment (nzw0301) [inline optuna/samplers/_tpe/sampler.py:717]: Without this change, the following warning message is shown:

```bash
/Users/nzw/Documents/optuna> pytest tests/samplers_tests/tpe_tests/test_sampler.py::test_split_observation_pairs_with_constraints_below_all_feasible
===================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
rootdir: /Users/nzw/Documents/optuna, configfile: pyproject.toml
collected 1 item

tests/samplers_tests/tpe_tests/test_sampler.py .                                                                                                                                        [100%]

====================================================================================== warnings summary =======================================================================================
tests/samplers_tests/tpe_tests/test_sampler.py::test_split_observation_pairs_with_constraints_below_all_feasible
  /opt/homebrew/Caskroom/miniconda/base/envs/optuna/lib/python3.9/site-packages/optuna-3.0.0.dev0-py3.9.egg/optuna/samplers/_tpe/sampler.py:717: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
    list(np.array(loss_vals)[feasible_idx]), n_below, None

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

File: tests/samplers_tests/test_grid.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def objective(trial: Trial) -> float:
     search_space = {
         "b": np.arange(-0.1, 0.1, 0.05),
         "c": ("x", "y", None, 1, 2.0),
-        "d": [-0.5, 0.5],
```
Comment (nzw0301) [inline tests/samplers_tests/test_grid.py:40]: These grid values are out-of-range in the corresponding suggest method, `trial.suggest_float("d", -5, 5, step=1)`.

File: tests/samplers_tests/test_partial_fixed.py [hunk 3]
```diff
@@ -52,7 +52,10 @@ def objective(trial: Trial) -> float:
         study.sampler = PartialFixedSampler(
             fixed_params={"y": fixed_y}, base_sampler=study.sampler
         )
-    study.optimize(objective, n_trials=1)
+    # Since `fixed_y` is out-of-the-range value in the corresponding suggest_int,
+    # `UserWarning` will occur.
+    with pytest.warns(UserWarning):
```
Comment (nzw0301) [inline tests/samplers_tests/test_partial_fixed.py:57]: Same as `test_out_of_the_range_numerical`'s `with` context.

File: tests/samplers_tests/test_grid.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def objective(trial: Trial) -> float:
...
+        "d": [5.0, 5.0],
```
Comment (toshihikoyanase) [inline tests/samplers_tests/test_grid.py:40]: I'm not sure, but do you intentionally set the same value (i.e., `5.0`)? Or is it a typo for `-5.0, 5.0`?

File: optuna/samplers/_tpe/sampler.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/samplers/_tpe/sampler.py:717]: The change looks reasonable. The current code `list(np.array(loss_vals)[feasible_idx])` returns `List[List[float, List[float]]]` but `_split_observation_pairs` expects `List[Tuple[float, List[float]]]`. So, I think this change also prevents future mypy errors.

**Review by toshihikoyanase — COMMENTED** [review]
The change basically looks good to me, but I have a question on `test_grid.py`.

**Comment by toshihikoyanase** [comment]
@contramundum53 Could you review this PR, please?

File: tests/samplers_tests/test_grid.py (hunk 4 above)
Comment (nzw0301) [reply tests/samplers_tests/test_grid.py]: This was a typo. Thank you for letting me know this. I've fixed it.

Comment (toshihikoyanase) [reply tests/samplers_tests/test_grid.py]: Thank you!

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM.

@HideakiImamura Do you think if we need to run benchmark before introducing this PR?

**Comment by contramundum53** [comment]
I ran the speed benchmark on MacBookAir(M1).

Master:
```
              ----------------------------- ------------
                  inmemory, random, 1000      193±8ms
                 inmemory, random, 10000     4.54±0.1s
                   inmemory, tpe, 1000       3.75±0.06s
                  inmemory, cmaes, 1000       548±8ms
                   sqlite, random, 1000      9.82±0.09s
               cached_sqlite, random, 1000   9.85±0.03s
                   redis, random, 1000       6.58±0.03s
                cached_redis, random, 1000   6.55±0.01s
              ============================= ============
```

This PR:
```
              ============================= ============
                storage, sampler, n_trials
              ----------------------------- ------------
                  inmemory, random, 1000      186±3ms
                 inmemory, random, 10000     4.43±0.2s
                   inmemory, tpe, 1000       3.76±0.03s
                  inmemory, cmaes, 1000       525±6ms
                   sqlite, random, 1000      11.0±0.6s
               cached_sqlite, random, 1000   10.4±0.4s
                   redis, random, 1000       6.58±0.02s
                cached_redis, random, 1000   6.52±0.02s
              ============================= ============
```
There is little performance regression. LGTM.

**Review by contramundum53 — APPROVED** [review]

**Pull request merged**